### DISCUSSION TITLE
Programming Atmel 29c256

### DISCUSSION THREAD

**Discussion by prlaba** [TomNisbet/TommyPROM#48] [Q&A]
I have some Atmel 29C256 flash chips I need to program.  Pinwise they're identical to the 28C256 chips, except the A14 and WE pins are swapped.

Based on my reading of its datasheet, the 29C256 writes uses a 64-byte buffer (Atmel calls it a page) for writing.

Other than swapping the two pins on the 29C256, will I need to make any changes to the Arduino software to program it?

By the way, I'm very happy to see that PCBs are now available.  I plan to order the 32-pin version, as most of the EEPROMs I need to program are 32 pins (29C020 and 29C040).

Thanks for a great product!

**Comment by TomNisbet** [comment]
Those chips should work with the default 28C code.  I haven't tried a 29C256, but have successfully burned a 29C010.  I just uploaded gerbers for an updated version of the TommyPROM32 board.  This has a second switchable VPP line and some other small changes.

Be aware that some 29C040 chips have a 256 byte sector size.  Those won't work with the current code because the 128 byte packet size in XModem will cause two separate write operations and the chip will erase the sector for each write.

**Reply by prlaba** [reply]
Thanks Tom!

Yes I saw your caution in the documentation re the 29c020 chips (I'll be programming some of those too, along with some 29c040's).  But I didn't see anything mentioned about the smaller 29c chips, like the 29c256, that's why I asked.

On a slightly different topic...

The TommyPROM programmer is a great alternative to the more expensive ones.  And now that TommyPROM PCBs are available-the 32-bit one is the way to go-it's starting to morph into a real product.

One feature that all the expensive programmers have is their ability to automatically detect the chip being programmed.  Most of them support hundreds (thousands?) of different chips.

The real advantage is that you simply plug your chip into the programmer's ZIF socket and the hardware/software does all the rest.

A modest improvement for the TommyPROM would be to assign IDs to a small set of supported EEPROMs, where each ID represented a specific routing of data and control signals to pins on a ZIF socket.  So the 28C256 and 29C256 would have separate IDs because they have slightly different pin outs.  Other chips might have different IDs because they have the same pin outs but different buffer sizes.

A set of ROMs would take as their input the n-bit ID of the chip to be programmed, and generate the correct routing of signals going to the ZIF's pins.  (A 32-pin ZIF would require five 2Kx8 ROMs, wired in parallel).  The ID would also instruct the software how to program the chip.

A user could manually set the ID of the chip to be programmed via an 8-bit (256 IDs) or 10-bit (1024 IDs) DIP switch pack.

New chips could be supported simply by assigning new IDs and reprogramming the ROMs.

A future version of the TommyPROM software could loop through all of the defined IDs, looking for a match on the pin routing.  How that would actually work is left as an exercise to the interested reader. 🙂

**Reply by TomNisbet** [reply]
That's an interesting idea, but it's well beyond the capabilities of the current design. There just isn't enough space on the Arduino to hold the UI, the programming algorithms, and the communications code. The existing design is modular in the sense that it can be recompiled to support different families, but it's not loading all of the possible code on the Arduino for each variant.  There's also no hardware to switch the signals around. 

This was really designed for one-off programming using parts that you probably already have laying around. I made the 28C256 PCB because I was programming a bunch of those for a project.  The more modular 32-pin board was because I've been experimenting with lots of different chips and wanted something a little cleaner than a breadboard because usually only a small number of signals change from chip to chip.

To design something like you are describing, it would make more sense to move all of the UI code to a program that runs on a host computer and then have the Arduino talk to the host with a lightweight communications protocol. That would let you have a GUI and a file system and get rid of the whole Modem thing.

**Reply by prlaba** [reply]
Thanks again, it helps me understand the design decisions you made.

Once you came up with the TommyPROM32 PCB, I started looking at your design differently.

Have you considered moving your hardware and software to a different microprocessor, like the Teensy?  You'd have more GPIO pins (no need for the shift registers), more room for the code (probably enough to store all of the different drivers), and much higher speed.  It would raise the cost of the hardware, but you'd have a better base for expanding your design, supporting more chips, etc.
